base/channels.jl

```diff
@@ -304,14 +427,54 @@ For unbuffered channels, blocks until a [`put!`](@ref) is performed by a differe
 task.
 """
 take!(c::Channel) = isbuffered(c) ? take_buffered(c) : take_unbuffered(c)
+
+if JULIA_PARTR
+
+function take_buffered(c::Channel)
+    while true
+        check_channel_state(c)
+        lock(c.lock)
+        if length(c.data) > 0
+            v = popfirst!(c.data)
+            unlock(c.lock)
+            notify(c.cond_put, nothing, false, false)
+            return v
+        end
+        unlock(c.lock)
+        wait(c.cond_take)
+    end
+end
+
+else # !JULIA_PARTR
+
 function take_buffered(c::Channel)
     wait(c)
     v = popfirst!(c.data)
     notify(c.cond_put, nothing, false, false) # notify only one, since only one slot has become available for a put!.
     v
 end
 
-popfirst!(c::Channel) = take!(c)
+end # !JULIA_PARTR
+
+if JULIA_PARTR
+
+function take_unbuffered(c::Channel{T}) where T
+    check_channel_state(c)
+    lock(c.lock)
+    push!(c.takers, current_task())
+    unlock(c.lock)
+    notify(c.cond_put, nothing, false, false)
+    try
+        return wait()::T
+    catch ex
+        lock(c.lock)
+        filter!(x->x!=current_task(), c.takers)
+        unlock(c.lock)
+        rethrow(ex)
+    end
+end
```

Comment:
The locks/unlocks in this section could be written in `lock(func)` form.
In 0.6 I found this had a small overhead over writing `lock` then `unlock` ,
but in theory it should just inline the anon func out of existence.

Reply:
As a side point, I like that the PATR code for this one is a lot clearer than the nonPATR code :-D